tools/shared/src/main/scala/org/scalajs/core/tools/linker/analyzer/Analyzer.scala

```diff
@@ -416,47 +416,55 @@ private final class Analyzer(semantics: Semantics,
-         * algorithm defining `java.lang.Class.getMethod`. This mimics how
-         * reflective calls are implemented on the JVM, at link time.
-         *
-         * Caveat: protected methods are not ignored. This can only make an
-         * otherwise invalid reflective call suddenly able to call a protected
-         * method. It never breaks valid reflective calls. This could be fixed
-         * if the IR retained the information that a method is protected.
-         */
-
-        @tailrec
-        def loop(ancestorInfo: ClassInfo): Option[MethodInfo] = {
-          if (ancestorInfo ne null) {
-            ancestorInfo.methodInfos.get(proxyName) match {
-              case Some(m) =>
-                assert(m.isReflProxy && !m.isAbstract)
-                Some(m)
-
-              case _ =>
-                ancestorInfo.findProxyMatch(proxyName) match {
-                  case Some(target) =>
-                    val targetName = target.encodedName
-                    Some(ancestorInfo.createReflProxy(proxyName, targetName))
-
-                  case None =>
-                    loop(ancestorInfo.superClass)
-                }
-            }
-          } else {
-            None
+        methodInfos.get(proxyName).orElse {
+          findReflectiveTarget(proxyName).map { reflectiveTarget =>
+            createReflProxy(proxyName, reflectiveTarget.encodedName)
           }
         }
+      }
+    }
+
+    private def findReflectiveTarget(proxyName: String): Option[MethodInfo] = {
+      /* The lookup for a target method in this code implements the
+       * algorithm defining `java.lang.Class.getMethod`. This mimics how
+       * reflective calls are implemented on the JVM, at link time.
+       *
+       * We add a bit of guess-work for default methods, as the documentation
+       * is very vague about them. Basically, we just take the first match in
+       * `ancestors`, as it's easy, and we're in a gray area anyway. At least,
+       * this will work when there is no overload.
+       *
+       * Caveat: protected methods are not ignored. This can only make an
+       * otherwise invalid reflective call suddenly able to call a protected
+       * method. It never breaks valid reflective calls. This could be fixed
+       * if the IR retained the information that a method is protected.
+       */
 
-        loop(this)
+      @tailrec
+      def loopParentClasses(ancestorInfo: ClassInfo): Option[MethodInfo] = {
+        if (ancestorInfo ne null) {
+          ancestorInfo.findProxyMatch(proxyName) match {
+            case Some(m) =>
+              Some(m)
+
+            case None =>
+              loopParentClasses(ancestorInfo.superClass)
+          }
+        } else {
+          // Look in default methods
+          ancestors.iterator.map(_.findProxyMatch(proxyName)).collectFirst {
+            case Some(m) => m
+          }
+        }
       }
+
+      loopParentClasses(this)
```

Comment:
How about:

```scala
val superClasses = Iterator.iterate(this)(_.superClass).takeWhile(_ ne null)
val candidates = superClasses.flatMap(_.findProxyMatch(proxyName)) ++
    ancestors.iterator.flatMap(_.findProxyMatch(proxyName))

candidates.headOption
```

Does the same (probably at the same speed as well) but doesn't mix two styles (tailrec and iterator) and is much shorter.